internal/security/bootstrapper/command/setupacl/aclpolicies.go

```diff
@@ -0,0 +1,197 @@
+	# HCL definition of server agent policy for EdgeX
+	agent "" {
+		policy = "read"
+	}
+	agent_prefix "edgex" {
+		policy = "write"
+	}
+	node "" {
+  		policy = "read"
+	}
+	node_prefix "edgex" {
+		policy = "write"
+	}
+	service "" {
+  		policy = "read"
+	}
+	service_prefix "edgex" {
+  		policy = "write"
+	}
+	# allow key value store put
+	# once the default_policy is switched to "deny",
+	# this is needed if wants to allow updating Key/Value configuration
+	key "" {
+  		policy = "read"
+	}
+	key_prefix "edgex" {
+		policy = "write"
+	}
+	`
+
+	// edgeXServicePolicyName is the name of the agent policy for edgex
+	edgeXServicePolicyName = "edgex-service-policy"
+
+	consulCreatePolicyAPI     = "/v1/acl/policy"
+	consulReadPolicyByNameAPI = "/v1/acl/policy/name/%s"
+
+	aclNotFoundMessage = "ACL not found"
+)
+
+// getOrCreateRegistryPolicy retrieves or creates a new policy
+// it inserts a new policy if the policy name does not exist and returns a policy
+// it returns the same policy if the policy name already exists
+func (c *cmd) getOrCreateRegistryPolicy(tokenID, policyName, policyRules string) (*Policy, error) {
+	// try to get the policy to see if it exists or not
+	policy, err := c.getPolicyByName(tokenID, policyName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get policy ID by name %s: %v", policyName, err)
+	}
+
+	if policy != nil {
+		// policy exists, return this one
+		return policy, nil
+	}
+
+	createPolicyURL, err := c.getRegistryApiUrl(consulCreatePolicyAPI)
+	if err != nil {
+		return nil, err
+	}
+
+	// payload struct for creating a new policy
+	type CreatePolicy struct {
+		Name        string `json:"Name"`
+		Description string `json:"Description,omitempty"`
+		Rules       string `json:"Rules,omitempty"`
+	}
+
+	createPolicy := &CreatePolicy{
+		Name:        policyName,
+		Description: "agent policy for EdgeX microservices",
+		Rules:       policyRules,
+	}
+
+	jsonPayload, err := json.Marshal(createPolicy)
+	c.loggingClient.Tracef("payload: %v", createPolicy)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal CreatePolicy JSON string payload: %v", err)
+	}
+
+	req, err := http.NewRequest(http.MethodPut, createPolicyURL, bytes.NewBuffer(jsonPayload))
+	if err != nil {
+		return nil, fmt.Errorf("failed to prepare create a new policy request for http URL: %w", err)
+	}
+
+	req.Header.Add(consulTokenHeader, tokenID)
+	req.Header.Add(clients.ContentType, clients.ContentTypeJSON)
+	resp, err := c.client.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to send create a new policy request for http URL: %w", err)
+	}
+
+	defer func() {
+		_ = resp.Body.Close()
+	}()
+
+	createPolicyResp, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to read create a new policy response body: %w", err)
+	}
+
+	var created Policy
+
+	switch resp.StatusCode {
+	case http.StatusOK:
+		if err := json.NewDecoder(bytes.NewReader(createPolicyResp)).Decode(&created); err != nil {
```

Comment:
Seems that it would make more sense to use NewDecoder(resp.Body) instead.

Reply:
never seen people using that way, resp.Body is also io.ReadCloser type not the io.Reader type; kind of different.
I think the byte.NewReader is de facto way

Reply:
Not true.  io.ReadCloser is defined as
```
type ReadCloser interface {
    Reader
    Closer
}
```

The difference is that ioutil.ReadAll is double-buffering the response -- reading it entirely into memory, and the putting another reader on top to process it.  It should be more memory efficient to just process it as a stream.

Reply:
if just for StatusOk case, ok with your suggestion but we have also other error cases like `StatusForbidden` and other status code, and in which case we want to get the error messages from response body so that is why the benefit to read it first into a [] byte variable works.
Please see how this `readPolicyResp` variable is used in other places as well.

```go
readPolicyResp, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("Failed to read readPolicyByName response body: %w", err)
	}

	fmt.Println("getPolicyByName call the response status code = ", resp.StatusCode)

	switch resp.StatusCode {
	case http.StatusOK:
		var existing Policy
		if err := json.NewDecoder(bytes.NewReader(readPolicyResp)).Decode(&existing); err != nil {
			return nil, fmt.Errorf("failed to decode Policy json data: %v", err)
		}

		return &existing, nil
	case http.StatusForbidden:
		// when the policy cannot be found by the name, the body returns "ACL not found"
		// so we treat it as non-error case
		if strings.EqualFold(aclNotFoundMessage, string(readPolicyResp)) {
			return nil, nil
		}

		return nil, fmt.Errorf("failed to read policy by name with error %s", string(readPolicyResp))
	default:
		return nil, fmt.Errorf("failed to read policy by name with status code= %d: %s",
			resp.StatusCode, string(readPolicyResp))
	}
```